libcontainer/container_linux.go

```diff
@@ -1124,6 +1133,16 @@ func (c *linuxContainer) runType() (Status, error) {
 		}
 		return Stopped, newSystemErrorWithCausef(err, "sending signal 0 to pid %d", pid)
 	}
+	// check if the init process is still the same process as the initial one,
+	// it could happen that the process is exited and a new process created with
+	// the same pid, in this case, the container would already be stopped.
+	startTime, err := system.GetProcessStartTime(pid)
+	if err != nil {
+		return Stopped, newSystemErrorWithCausef(err, "getting init process %d start time", pid)
+	}
+	if c.initProcessStartTime != startTime {
+		return Stopped, nil
```

Comment:
This was my suggestion, but @cyphar suggested [stat'ing `/proc/{pid}`][1] (and using the inode?).  In either case, it may be better to abstract this extra identifier out into a `GetProcessUniquifier` helper (silly name, but whatever ;), since we're only using it for an equality comparison, and don't really care what the value actually means.

[1]: https://github.com/opencontainers/runc/issues/871#issuecomment-223636697